source/opt/ssa_rewrite_pass.cpp

```diff
@@ -0,0 +1,577 @@
+
+// This file implements the SSA rewriting algorithm proposed in
+//
+//      Simple and Efficient Construction of Static Single Assignment Form.
+//      Braun M., Buchwald S., Hack S., Leißa R., Mallon C., Zwinkau A. (2013)
+//      In: Jhala R., De Bosschere K. (eds)
+//      Compiler Construction. CC 2013.
+//      Lecture Notes in Computer Science, vol 7791.
+//      Springer, Berlin, Heidelberg
+//
+//      https://link.springer.com/chapter/10.1007/978-3-642-37051-9_6
+//
+// In contrast to common eager algorithms based on dominance and dominance
+// frontier information, this algorithm works backwards from load operations.
+//
+// When a target variable is loaded, it queries the variable's reaching
+// definition.  If the reaching definition is unknown at the current location,
+// it searches backwards in the CFG, inserting Phi instructions at join points
+// in the CFG along the way until it finds the desired store instruction.
+//
+// The algorithm avoids repeated lookups using memoization.
+//
+// For reducible CFGs, which are a superset of the structured CFGs in SPIRV,
+// this algorithm is proven to produce minimal SSA.  That is, it inserts the
+// minimal number of Phi instructions required to ensure the SSA property, but
+// some Phi instructions may be dead
+// (https://en.wikipedia.org/wiki/Static_single_assignment_form).
+
+#include "ssa_rewrite_pass.h"
+#include "cfg.h"
+#include "make_unique.h"
+#include "mem_pass.h"
+#include "opcode.h"
+
+#include <sstream>
+
+// Debug logging (0: Off, 1-N: Verbosity level).  Replace this with the
+// implementation done for
+// https://github.com/KhronosGroup/SPIRV-Tools/issues/1351
+// #define SSA_REWRITE_DEBUGGING_LEVEL 1
+
+#ifdef SSA_REWRITE_DEBUGGING_LEVEL
+#include <ostream>
+#else
+#define SSA_REWRITE_DEBUGGING_LEVEL 0
+#endif
+
+namespace spvtools {
+namespace opt {
+
+namespace {
+const uint32_t kStoreValIdInIdx = 1;
+const uint32_t kVariableInitIdInIdx = 1;
+}  // namespace
+
+std::string SSARewriter::PhiCandidate::PrettyPrint(const ir::CFG* cfg) const {
+  std::ostringstream str;
+  str << "%" << result_id_ << " = Phi[%" << var_id_ << ", BB %" << bb_->id()
+      << "](";
+  if (phi_args_.size() > 0) {
+    uint32_t arg_ix = 0;
+    for (uint32_t pred_label : cfg->preds(bb_->id())) {
+      uint32_t arg_id = phi_args_[arg_ix++];
+      str << "[%" << arg_id << ", bb(%" << pred_label << ")] ";
+    }
+  }
+  str << ")" << ((is_trivial_) ? " [TRIVIAL]" : "")
+      << ((is_complete_) ? "  [COMPLETE]" : "  [INCOMPLETE]");
+
+  return str.str();
+}
+
+SSARewriter::PhiCandidate& SSARewriter::CreatePhiCandidate(uint32_t var_id,
+                                                           ir::BasicBlock* bb) {
+  uint32_t phi_result_id = pass_->context()->TakeNextId();
+  auto result = phi_candidates_.emplace(
+      phi_result_id, PhiCandidate(var_id, phi_result_id, bb));
+
+  // We should never try to create more than one Phi candidate for the same
+  // |var_id| on |bb|.
```

Comment:
The whole pass is needed because we access services other than the context.  In a subsequent PR, I'll see about moving those services into the rewriter, if they're not used anywhere else.  I did not want to add too much noise in this PR (it's already pretty big).

Reply:
Good point about the check.  Yeah, that check is now useless.  Thanks.